### PR TITLE
[USER32] Fix F1'97 Demo icon not showing in explorer

### PR DESCRIPTION
**Pull request by Doug-Lyons**
## Purpose

_Fix the F1'97 Demo program not showing an icon in explorer.._

JIRA issue: [CORE-10726](https://jira.reactos.org/browse/CORE-10726)

## Proposed changes

_If the normal method to get an icon fails, then provide an alternate one._

After patch:
CORE-10726-OK

Icon shown in W2K3SP2:

FI97_Icons_In_W2K3SP2

### CODE REVIEW

File: win32ss/user/user32/misc/exticon.c [hunk 1]
```diff
@@ -714,9 +714,69 @@ static UINT ICO_ExtractIconExW(
...
+                  BITMAPINFOHEADER bmih;
+                  RtlCopyMemory(&bmih, idata, sizeof(BITMAPINFOHEADER));
+                  /* Do the Width and Height look correct for an icon */
+                  if (bmih.biWidth == 32 && bmih.biHeight == 64)
+                  {
+                      RetPtr[0] = CreateIconFromResourceEx(idata, igdataent->Size,
+                          TRUE, 0x00030000, cx1, cy1, flags);
+                      if (cx2 && cy2)
+                          RetPtr[1] = CreateIconFromResourceEx(idata, idataent->Size,
+                              TRUE, 0x00030000, cx2, cy2, flags);
```
Comment (HBelusca) [inline win32ss/user/user32/misc/exticon.c:770]: What is this magic 0x00030000 value?

Comment (Doug-Lyons) [reply win32ss/user/user32/misc/exticon.c:771]: [in] dwVer

Type: DWORD

The version number of the icon or cursor format for the resource bits pointed to by the pbIconBits parameter. The value must be greater than or equal to 0x00020000 and less than or equal to 0x00030000. This parameter is generally set to 0x00030000.

File: win32ss/user/user32/misc/exticon.c [hunk 2]
```diff
@@ -714,9 +714,75 @@ static UINT ICO_ExtractIconExW(
...
+                  continue;
+              }
+
+              /* Check to see if this looks like an icon bitmap */
+              if (idata[0] == sizeof(BITMAPINFOHEADER))
+              {
+                  BITMAPINFOHEADER bmih;
+                  RtlCopyMemory(&bmih, idata, sizeof(BITMAPINFOHEADER));
+                  /* Do the Width and Height look correct for an icon */
+                  if (bmih.biWidth == 32 && bmih.biHeight == 64)
```
Comment (KRosUser2) [inline win32ss/user/user32/misc/exticon.c:765]: Is a valid icon 32 x 64 only ?
Is it really a && ? Or || using both dimensions in check ?

Comment (Doug-Lyons) [reply win32ss/user/user32/misc/exticon.c:765]: This checks for a valid 32x32 icon only. The height is twice as much because there is one full sized map for the AND icon (actual image) and another full sized map for the XOR (transparency map). We want to check for both of these sizes being correct before we accept the icon as valid. Thanks.

Comment (katahiromz) [reply win32ss/user/user32/misc/exticon.c:765]:
```suggestion
                  if ((bmih.biWidth * 2) == bmih.biHeight)
```

Comment (Doug-Lyons) [reply win32ss/user/user32/misc/exticon.c]: Good Idea. Done. Thanks.

File: win32ss/user/user32/misc/exticon.c [hunk 3]
```diff
@@ -714,9 +714,75 @@ static UINT ICO_ExtractIconExW(
...
+
+              /* Try to get an icon by walking the files Resource Section */
+              if (iconresdir->NumberOfIdEntries > 0)
+              {
+                  xresent = (const IMAGE_RESOURCE_DIRECTORY_ENTRY*)
+                      (ULONG_PTR)(iconresdir + 1);
+              }
+              else
+              {
+                  RetPtr[i]=0;
```
Comment (katahiromz) [inline win32ss/user/user32/misc/exticon.c:729]:
```suggestion
                  RetPtr[i] = 0;
```

File: win32ss/user/user32/misc/exticon.c [hunk 4]
```diff
@@ -714,9 +714,75 @@ static UINT ICO_ExtractIconExW(
...
+                       ((const char *)rootresdir + xresent->OffsetToDirectory);
+
+              if (resdir->NumberOfIdEntries > 0) // Do we have entries
+              {
+                  xresent = (const IMAGE_RESOURCE_DIRECTORY_ENTRY*)
+                      (ULONG_PTR)(resdir + 1);
+              }
+              else
+              {
+                  RetPtr[i]=0;
```
Comment (katahiromz) [inline win32ss/user/user32/misc/exticon.c:744]:
```suggestion
                  RetPtr[i] = 0;
```

File: win32ss/user/user32/misc/exticon.c [hunk 5]
```diff
@@ -714,9 +714,75 @@ static UINT ICO_ExtractIconExW(
...
+              }
+
+              /* Retreive the data entry and find its address */
+              igdataent = (const IMAGE_RESOURCE_DATA_ENTRY*)
+                  ((const char *)rootresdir + xresent->OffsetToData);
+              idata = RtlImageRvaToVa(RtlImageNtHeader((HMODULE)peimage),
+                  (HMODULE)peimage, igdataent->OffsetToData, NULL);
+              if (!idata)
+              {
+                  RetPtr[i]=0;
```
Comment (katahiromz) [inline win32ss/user/user32/misc/exticon.c:755]:
```suggestion
                  RetPtr[i] = 0;
```

File: win32ss/user/user32/misc/exticon.c [hunk 6]
```diff
@@ -714,9 +714,75 @@ static UINT ICO_ExtractIconExW(
...
+                      if (cx2 && cy2)
+                          RetPtr[1] = CreateIconFromResourceEx(idata, idataent->Size,
+                              TRUE, 0x00030000, cx2, cy2, flags);
+                      ret = 1;  // Set number of icons found
+                      goto end; // Success so Exit
+                  }
+              }
+              else
+              {
+                  RetPtr[i]=0;
```
Comment (katahiromz) [inline win32ss/user/user32/misc/exticon.c:778]:
```suggestion
                  RetPtr[i] = 0;
```

File: win32ss/user/user32/misc/exticon.c [hunk 7]
```diff
@@ -714,9 +714,75 @@ static UINT ICO_ExtractIconExW(
...
+                  xresent = (const IMAGE_RESOURCE_DIRECTORY_ENTRY*)
+                      (ULONG_PTR)(resdir + 1);
+              }
+              else
+              {
+                  RetPtr[i]=0;
+                  continue;
+              }
+
+              /* Retreive the data entry and find its address */
```
Comment (JoachimHenze) [inline win32ss/user/user32/misc/exticon.c:748]:
```suggestion
              /* Retrieve the data entry and find its address */
```

File: win32ss/user/user32/misc/exticon.c (hunk 3 above)
Comment (Doug-Lyons) [reply win32ss/user/user32/misc/exticon.c]: Fixed. Thanks.

File: win32ss/user/user32/misc/exticon.c (hunk 4 above)
Comment (Doug-Lyons) [reply win32ss/user/user32/misc/exticon.c]: Fixed. Thanks.

File: win32ss/user/user32/misc/exticon.c (hunk 5 above)
Comment (Doug-Lyons) [reply win32ss/user/user32/misc/exticon.c]: Fixed. Thanks.

File: win32ss/user/user32/misc/exticon.c (hunk 6 above)
Comment (Doug-Lyons) [reply win32ss/user/user32/misc/exticon.c]: Fixed. Thanks.

File: win32ss/user/user32/misc/exticon.c (hunk 7 above)
Comment (Doug-Lyons) [reply win32ss/user/user32/misc/exticon.c]: Fixed. Thanks.

**Review by KRosUser2 — COMMENTED** [review]
no comment.

**Comment by jeditobe** [comment]
is it correct enogh  to merge?

**Comment by KRosUser2** [comment]
No adverse comment + testbot positively ran by @Doug-Lyons

**Review by HBelusca — APPROVED** [review]
I don't have much of an opinion on this. I just observe (not related to your changes in particular, but more an impression about that function) that a lot of small parts are repetitive...

File: win32ss/user/user32/misc/exticon.c [hunk 8]
```diff
@@ -714,9 +714,75 @@ static UINT ICO_ExtractIconExW(
...
+              else
+              {
+                  RetPtr[i] = 0;
+                  continue;
+              }
+#else
               WARN("icon entry %d not found\n", LOWORD(pIconId[i]));
 	      RetPtr[i]=0;
 	      continue;
+#endif
```
Comment (binarymaster) [inline win32ss/user/user32/misc/exticon.c:785]: What I can see here is looking more like a workaround. It's a big code block added at the failure case when the icon is not found. I think there should be a smaller proper fix targetting the code above, or the functions it uses, effectively making it not fail in this case.

Comment (JoachimHenze) [reply win32ss/user/user32/misc/exticon.c:785]: Yes, I also think it is a workaround for a (rare) corner case.
Suggestion:
1.) Add a 2k3sp2 pic at least that proves our target displaying that specific icon.
2.) Since it is likely that we will change this code again in the future due to it being a workaround, it would be good to have an automated test on the testbots which covers this case.
If both are met, I will give my Approval for committing that workaround.

Comment (Doug-Lyons) [reply win32ss/user/user32/misc/exticon.c:785]: @binarymaster and @JoachimHenze, each of you are certainly entitled to your opinion on whether this is a workaround. But, after working with this for quite some hours, I disagree. If you look at this file in CFF Explorer, which I have provided below, you will see that to check for the correctness of the 'Icon Group' tree, you must navigate the lower boxed area shown. Once you find that this contains inconsistent data, to get to the 'Icons' you must navigate the upper boxed area as shown below. Thanks very much to each of you for reviewing and providing input here.

F197_Resource_Directory

Comment (JoachimHenze) [reply win32ss/user/user32/misc/exticon.c:785]: We do better understand now.
The specific F197I.exe built with the WatCom C++ compiler is most likely not entirely legit.
Due to your further analysis we do know now that:
WinXP and 2k3sp2 do manage to display that icon in their explorer,
while Win7 shows only a blank icon.
I am aware that you wrote a testcase for that scenario, where you put all the information into also within a comment.
I will approve committing your test-case, and I will then afterwards also approve your fix.
Due to it being such a tricky corner case, I want to thank you again for having invested so much love in the analysis, in the test and in the fix.

Comment (Doug-Lyons) [reply win32ss/user/user32/misc/exticon.c:785]: Thanks for the positive comments. You are very welcome. I enjoy the challenges.

File: win32ss/user/user32/misc/exticon.c [hunk 9]
```diff
@@ -714,9 +714,75 @@ static UINT ICO_ExtractIconExW(
...
+              /* Try to get an icon by walking the files Resource Section */
+              if (iconresdir->NumberOfIdEntries > 0)
+              {
+                  xresent = (const IMAGE_RESOURCE_DIRECTORY_ENTRY*)
+                      (ULONG_PTR)(iconresdir + 1);
+              }
+              else
+              {
+                  RetPtr[i] = 0;
+                  continue;
```
Comment (SergeGautherie) [inline win32ss/user/user32/misc/exticon.c:730]: Create/Use a `goto`, instead of duplicating existing code.

File: win32ss/user/user32/misc/exticon.c [hunk 10]
```diff
@@ -714,9 +714,75 @@ static UINT ICO_ExtractIconExW(
 	    xresdir = find_entry_by_id(iconresdir, LOWORD(pIconId[i]), rootresdir);
             if( !xresdir )
             {
+#ifdef __REACTOS__
```
[the remark this answers is not in the document]
Comment (Doug-Lyons) [reply win32ss/user/user32/misc/exticon.c:717]: Done. Thanks.

File: win32ss/user/user32/misc/exticon.c (hunk 9 above)
Comment (Doug-Lyons) [reply win32ss/user/user32/misc/exticon.c:733]: No thanks. Maybe you can do another PR later for this.

Comment (SergeGautherie) [reply win32ss/user/user32/misc/exticon.c:733]: Hum :-|

File: win32ss/user/user32/misc/exticon.c [hunk 11]
```diff
@@ -714,9 +714,78 @@ static UINT ICO_ExtractIconExW(
...
+              /* XP/2K3 can decode icons this way. Vista/Win7, cannot. This
+               * handles damaged Resources in 'Icon Group' by using the 'Icon'
+               * resources. Error can occur in older Watcom C/C++ compiles. */
+              const IMAGE_RESOURCE_DIRECTORY *resdir;
+              WARN("icon entry %d not found\n", LOWORD(pIconId[i]));
```
[the remark this answers is not in the document]
Comment (Doug-Lyons) [reply win32ss/user/user32/misc/exticon.c:722]: It would only be necessary to someone to look at the source of the WARN in the code and read the comments above to know this. So I consider it as not needed. Thanks.

File: win32ss/user/user32/misc/exticon.c [hunk 12]
```diff
@@ -714,9 +714,78 @@ static UINT ICO_ExtractIconExW(
 	    xresdir = find_entry_by_id(iconresdir, LOWORD(pIconId[i]), rootresdir);
             if( !xresdir )
             {
+#ifdef __REACTOS__
+              /* XP/2K3 can decode icons this way. Vista/Win7, cannot. This
+               * handles damaged Resources in 'Icon Group' by using the 'Icon'
+               * resources. Error can occur in older Watcom C/C++ compiles. */
```
[the remark this answers is not in the document]
Comment (Doug-Lyons) [reply win32ss/user/user32/misc/exticon.c:720]: I will try and do this update if other changes are needed. Thanks.

[the remark this answers is not in the document]
Comment (JoachimHenze) [reply win32ss/user/user32/misc/exticon.c:720]: A good refinement proposed by Serge. I will approve the current state already, but I do think you should accept his suggestion.

**Review by JoachimHenze — APPROVED** [review]

File: win32ss/user/user32/misc/exticon.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (Doug-Lyons) [reply win32ss/user/user32/misc/exticon.c]: @JoachimHenze, I was going to update this, but it looks like you were faster than I was. Thanks.

**Comment by Doug-Lyons** [comment]
Thanks, @HBelusca and @JoachimHenze for your help with approvals and committing.
